Integrations/integration-Tufin.yml

```diff
@@ -0,0 +1,321 @@
+
+        # Build the query from user input
+        querystring = {
+            "src": demisto.args()['Source'],
+            "dst": demisto.args()['Destination'],
+            "service": demisto.args().get('Service', 'Any'),
+            "includeIncompletePaths": "true"
+        }
+
+        # Define the basic output for the function, augmenting later with TOS data
+        entry = {
+            'Type': entryTypes['note'],
+            'Contents': '',
+            'ContentsFormat': formats['json'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': '',
+            'EntryContext': {}
+        }
+
+        # Ask TOS for the path
+        o = tos_request("/securetrack/api/topology/path", querystring)
+
+        # Verify the data and return
+        try:
+            entry['EntryContext']['Tufin.TrafficAllowed'] = o['path_calc_results']['traffic_allowed']
+            entry['EntryContext']['Tufin.TrafficRawJSON'] = o['path_calc_results']['device_info']
+            entry['EntryContext']['Tufin.TrafficDevices'] = [d['name'] for d in o['path_calc_results']['device_info']]
+            entry['Contents'] = o['path_calc_results']['device_info']
+            entry['HumanReadable'] = tableToMarkdown('Tufin Topology Search for {} to {} via Service {}. Traffic is {}'.format(querystring['src'], querystring['dst'], querystring['service'], ('**Denied**', '**Allowed**')[o['path_calc_results']['traffic_allowed']]),
+            {'Start': querystring['src'], 'Devices in Path' : "-->".join(['**' + d['name'] + '**' + ' ({})'.format(d['vendor']) for d in o['path_calc_results']['device_info']]), 'End': querystring['dst']}, ['Start', 'Devices in Path', 'End'])
+        except KeyError:
+            entry['HumanReadable'] = 'Unknown Output Returned'
+            entry['Contents'] = 'Unknown Output Returned'
+            entry['ContentsFormat'] = formats['text']
+            entry['Type'] = entryTypes['error']
+            return entry
+
+        return entry
+
+
+    def path_finder_image():
+        """ Sample query: querystring = {"src":"10.80.80.0","dst":"172.16.200.80","service":"tcp:80","includeIncompletePaths":"true","displayBlockedStatus":"true"} """
+        querystring = {
+            "src": demisto.args()['Source'],
+            "dst": demisto.args()['Destination'],
+            "service": demisto.args().get('Service', 'Any'),
+            "includeIncompletePaths": "true",
+            "displayBlockedStatus": "true"
+        }
+
+        entry = {
+            'Type': entryTypes['note'],
+            'Contents': '',
+            'ContentsFormat': formats['text'],
+            'ReadableContentsFormat': formats['markdown'],
+            'HumanReadable': ''
+        }
+
+        try:
+            img = tos_request("/securetrack/api/topology/path_image", querystring, {'accept': "image/png",'content-type': "application/json",'cache-control': "no-cache"})
+            if len(img) > 100:
```

Comment:
Could you please add a comment to explain the logic here ?

Reply:
next commit